prepration for running extension in nodeless environment

rebornix opened the pull request:
<!--
Yay! Thanks for sending us a PR! 🎊

Please ensure your PR adheres to:

- [x] Commit messages has a short & issue references when necessary
- [x] Each commit does a logical chunk of work.
- [x] It builds and tests pass (e.g `gulp`)
-->

**What this PR does / why we need it**:

VS Code is adding support for running extensions in nodeless environment. This PR tries to refactor the code a bit to allow later on replacing node APIs (`fs`, `child_process`) with shims. 

This PR should not change any functionality for running Vim in desktop.

**Which issue(s) this PR fixes**

Once we can run Vim in web worker, we can also think about running Vim in its own worker, which at the end will help with the performance issue: [Overriding the default 'type' command and then calling the default 'type' command results in significantly slower execution time ](https://github.com/microsoft/vscode/issues/65876)

<!--
Commits in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)
-->

**Special notes for your reviewer**:

J-Fields commented on src/configuration/vimrc.ts:
```diff
@@ -50,12 +50,16 @@ class VimrcImpl {
+            VimrcImpl.addRemapToConfig(config, remap);
+          }
         }
+      } finally {
```
Shouldn't this be `catch`?

J-Fields commented on src/state/vimState.ts:
```diff
@@ -25,6 +24,7 @@ import { ReplaceState } from './../state/replaceState';
  */
 export class VimState implements vscode.Disposable {
   private readonly logger = Logger.get('VimState');
+  private _disposables: vscode.Disposable[] = [];
```
Is this needed if nothing `VimState` holds is disposable (after you pulled `nvim` out)?

J-Fields commented on extension.ts:
```diff
@@ -3,528 +3,27 @@
-
-  return curHandler;
-}
+import './src/actions/include-main';
```
What's the rationale for splitting this file into `extension.ts` and `extensionBase.ts`?

rebornix replied:
The names are not intuitive and at the beginning I call them `extensionNodeEntry.ts`, `extensionBase.ts` and then we later on we introduce `extensionWebEntry.ts`. The idea is that we will have different entry files for node and web, and the major difference between the entry files are the modules they load and the default flag/options based on the environment. The `extensionBase` will be platform/env agnostic ultimately.

 I didn't make it so now as there are quite a few places (package.json, test, etc) rely on `extension.js` so I try to keep the change minimal.

J-Fields replied:
Got it 👍

J-Fields commented on src/cmd_line/commandLine.ts:
```diff
@@ -33,13 +39,17 @@ class CommandLine {
-  public async load(): Promise<void> {
+  public async load(context: vscode.ExtensionContext): Promise<void> {
+    const m = await import('../neovim/neovim');
+    this._nvim = new m.NeovimWrapper();
```
Is this getting disposed anywhere?

rebornix commented:
@J-Fields addressed your feedback, thanks for them!

Based on our code structure, to allow us ship bundles for different platforms (node/web), we need to do multiple different refactorings (not all included in this PR)

* static contributions if possible. `Actions` are the first set of modules we made static, each `*action` module is *plug-and-play*. In this PR, I made configuration validators static as well so in Web, all we need to do is not including `import './src/configuration/validators/neovimValidator';` to get validators work.
* Make `extensionBase` platform agnostic if possible. For example, in Web, `context.globalStoragePath` might not be available so we should not depend on it directly. We can move most platform specific configurations into entry files, e.g., `extension.ts`, `extensionWeb.ts`
* Lastly, and also the dirty part. We have a lot of code directly calling `fs`, `child_process`, we need to carefully shim them. The work behind the this will come in another PR.

rebornix commented:
@J-Fields thanks for merging it with master ;)
